dokka-runners/dokka-gradle-plugin/src/main/kotlin/internal/PluginFeaturesService.kt

```diff
@@ -0,0 +1,144 @@
+/*
+ * Copyright 2014-2024 JetBrains s.r.o. Use of this source code is governed by the Apache 2.0 license.
+ */
+package org.jetbrains.dokka.gradle.internal
+
+import org.gradle.api.Project
+import org.gradle.api.logging.Logging
+import org.gradle.api.provider.Property
+import org.gradle.api.services.BuildService
+import org.gradle.api.services.BuildServiceParameters
+import org.gradle.kotlin.dsl.extra
+import org.gradle.kotlin.dsl.registerIfAbsent
+
+/**
+ * Internal utility service for managing Dokka Plugin features and warnings.
+ *
+ * Using a [BuildService] is most useful for only logging a single warning for the whole project,
+ * regardless of how many subprojects have applied DGP.
+ */
+internal abstract class PluginFeaturesService : BuildService<PluginFeaturesService.Params> {
+
+    interface Params : BuildServiceParameters {
+        /**
+         * Whether DGP should use V2 [org.jetbrains.dokka.gradle.DokkaBasePlugin].
+         *
+         * Otherwise, fallback to V1 [org.jetbrains.dokka.gradle.DokkaClassicPlugin].
+         */
+        val enableV2Plugin: Property<Boolean>
+
+        /** If `true`, suppress [PluginFeaturesService.logV2Message]. */
+        val suppressV2PluginMessage: Property<Boolean>
+    }
+
+    internal val enableV2Plugin: Boolean get() = parameters.enableV2Plugin.get()
+    private val suppressV2PluginMessage: Boolean get() = parameters.suppressV2PluginMessage.get()
+
+    /** Used to only log the V1 message once per project, regardless of how many subprojects there are. */
+    private var v1WarningLogged: Boolean = false
+
+    /** Used to only log the V2 message once per project, regardless of how many subprojects there are. */
+    private var v2MessageLogged: Boolean = false
+
+    fun logV1Warning() {
```

Comment:
I'm going to update the 'log only once' mechanism in https://github.com/Kotlin/dokka/pull/3743 to use `lazy {}`, which will make it easier to manage and also synchronised.

Reply:
I implemented the `lazy {}` approach in this PR as well, since I had to do a lot of refactoring to work around https://github.com/gradle/gradle/issues/17559.